confluent_kafka/schema_registry/avro.py

```diff
@@ -0,0 +1,329 @@
+
+        with _ContextStringIO() as fo:
+            # Write the magic byte and schema ID in network byte order (big endian)
+            fo.write(pack('>bI', _MAGIC_BYTE, self.schema_id))
+            # write the record to the rest of the buffer
+            schemaless_writer(fo, self.parsed_schema, obj)
+
+            return fo.getvalue()
+
+
+class AvroDeserializer(Deserializer):
+    """
+    AvroDeserializer decodes binary blobs written in the Schema Registry
+    binary format for Avro to an object.
+
+    Note:
+        ``Complex Types`` are returned as dicts. If a more specific instance
+        type is desired a callable , ``to_object``, may be registered with
+        the AvroDeserializer which converts a dict to the desired type.
+
+        See ``avroconsumer.py`` in the examples directory for example usage.
+
+    Args:
+        schema_registry_client (SchemaRegistryClient): Confluent Schema Registry
+            client instance.
+
+        schema_str (str): Avro reader schema declaration.
+
+        from_dict (callable): Callable(dict, SerializationContext) -> object.
+            Converts dict to an instance of some object.
+
+    .. _Schema declaration:
+        https://avro.apache.org/docs/current/spec.html#schemas
+
+    .. _Schema Resolution
+        https://avro.apache.org/docs/1.8.2/spec.html#Schema+Resolution
+
+    """
+    __slots__ = ['reader_schema', 'registry', 'from_dict', 'writer_schemas']
+
+    def __init__(self, schema_registry_client, schema_str, from_dict=None):
+        self.registry = schema_registry_client
+        self.writer_schemas = {}
+
+        self.reader_schema = parse_schema(loads(schema_str))
+
+        if from_dict is not None and not callable(from_dict):
+            raise ValueError("to_object must be callable with the signature"
+                             " to_object(SerializationContext, dict) -> object")
+        self.from_dict = from_dict
+
+    def __call__(self, value, ctx):
+        """
+        Decodes a Confluent Schema Registry Avro formatted binary blob
+        to an object.
+
+        Arguments:
+            value (bytes): binary blob
+
+            ctx (SerializationContext): Metadata pertaining to the serialization
+                operation.
+
+        Raises:
+            SerializerError if an error occurs ready data.
+
+        Returns:
+            dict: decoded object literal if value was not None, otherwise None
+
+        """
+        if value is None:
+            return None
+
+        if len(value) <= 5:
+            raise SerializationError("Message too small. This message was not"
+                                     " produced with a Confluent"
+                                     " Schema Registry serializer.")
+
+        with _ContextStringIO(value) as payload:
+            magic, schema_id = unpack('>bI', payload.read(5))
+            if magic != _MAGIC_BYTE:
+                raise SerializationError("Unknown magic byte. This message was"
```

Comment:
technically, the second sentence may not be true in the future, because the magic byte kind of doubles as a version number (which has never been updated).